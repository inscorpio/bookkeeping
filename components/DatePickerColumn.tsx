import useTouchMove from '~/hooks/useTouchMove'

interface Props {
  columns: string[]
  selectIndex: number
  onSelect: (selectIndex: number) => void
}

export const ITEM_HEIGHT = 28

// !!! 注意 translateY 是负数
export default function DatePickerColumn({ columns, selectIndex, onSelect }: Props) {
  const initialMovedIstance = selectIndex * ITEM_HEIGHT
  const { moveDistance, touchStart, touchMove } = useTouchMove(selectIndex, columns.length - 1)
  const finalDistance = initialMovedIstance + moveDistance
  const handleTouchEnd = () => {
    onSelect(finalDistance / ITEM_HEIGHT)
  }

  return (
    <li
      className="flex flex-col items-center overflow-hidden"
      onTouchStart={touchStart}
      onTouchMove={touchMove}
      onTouchEnd={handleTouchEnd}
    >
      {
        columns.map((v, i) => {
          return (
            <span
              key={`${v}-${i}`}
              style={{ transform: `translateY(${(finalDistance) * -1}px)` }}
            >
              {v}
            </span>
          )
        })
      }
    </li>
  )
}
