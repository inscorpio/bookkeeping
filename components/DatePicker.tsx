'use client'
import dayjs from 'dayjs'
import DatePickerColumn from '~/components/DatePickerColumn'
import Dialog from '~/components/Dialog'

interface Props {
  visible?: boolean
  value?: Date
  startDate?: number
  endDate?: number
  onClose: () => void
  onComfirm: (date: Date) => void
}

const date = new Date()

export default function DatePicker(
  {
    visible,
    value = date,
    startDate = dayjs(date).year() - 10,
    endDate = dayjs(date).year() + 10,
    onClose,
    onComfirm,
  }: Props,
) {
  const year = dayjs(value).year()
  const month = dayjs(value).month() + 1
  const day = dayjs(value).date()

  const years = Array.from({ length: endDate - startDate + 1 }).map((_, i) => `${startDate + i}年`)
  const months = Array.from({ length: 12 }).map((_, i) => `${i + 1}月`)
  const days = Array.from({ length: dayjs(value).daysInMonth() }).map((_, i) => `${i + 1}日`)

  let selectYearIndex = years.findIndex(v => v === `${year}年`)
  let selectMonthIndex = months.findIndex(v => v === `${month}月`)
  let selectDayIndex = days.findIndex(v => v === `${day}日`)

  const confirm = (
    <button onClick={() => {
      const year = years[selectYearIndex].slice(0, 4)
      const rawMonth = months[selectMonthIndex]
      const month = (rawMonth.length === 2
        ? `0${rawMonth}`
        : rawMonth
      ).slice(0, 2)
      const rawDay = days[selectDayIndex]
      const day = (rawDay.length === 2
        ? `0${rawDay}`
        : rawDay
      ).slice(0, 2)
      const dayJsDate = dayjs(`${year}-${month}-${day}`)
      const date = dayJsDate.toDate()
      onComfirm(date)
      onClose()
    }}
    >
      确定
    </button>
  )
  return (
    <Dialog visible={visible} confirm={confirm} onClose={onClose}>
      <div className="reactive h-full">
        <div className="absolute w-full h-7 border-y border-stone-300"></div>
        <ul className="flex justify-center gap-6 h-full text-lg">
          <DatePickerColumn
            columns={years}
            selectIndex={selectYearIndex}
            onSelect={(index) => {
              selectYearIndex = index
            }}
          />
          <DatePickerColumn
            columns={months}
            selectIndex={selectMonthIndex}
            onSelect={(index) => {
              selectMonthIndex = index
            }}
          />
          <DatePickerColumn
            columns={days}
            selectIndex={selectDayIndex}
            onSelect={(index) => {
              selectDayIndex = index
            }}
          />
        </ul>
      </div>
    </Dialog>
  )
}
