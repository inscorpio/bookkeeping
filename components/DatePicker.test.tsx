import { fireEvent, render, screen } from '@testing-library/react'
import dayjs from 'dayjs'
import { expect, test, vi } from 'vitest'
import DatePicker from '~/components/DatePicker'
import { touchMoveUp } from '~/components/DatePickerColumn.helper'

test('render', () => {
  const { container } = render(
    <DatePicker
      visible={true}
      value={new Date('2023-11-02')}
      onClose={() => {}}
      onComfirm={() => {}}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('onClose', async () => {
  let visible = true

  render(
    <DatePicker
      visible={visible}
      onClose={() => {
        visible = false
      }}
      onComfirm={() => {}}
    />,
  )

  const cancelButton = screen.getByRole('button', { name: '取消' })
  await fireEvent.click(cancelButton)
  expect(visible).toBe(false)
})

test('onComfirm when leading zero', async () => {
  let dummy
  const handleClose = vi.fn()
  render(
    <DatePicker
      visible={true}
      value={dayjs('2023-01-01').toDate()}
      onClose={handleClose}
      onComfirm={(value) => {
        dummy = dayjs(value).format('YYYY-MM-DD')
      }}
    />,
  )

  const [year, month, day] = screen.getAllByRole('listitem')
  await touchMoveUp(year)
  await touchMoveUp(month)
  await touchMoveUp(day)
  const confirmButton = screen.getByRole('button', { name: '确定' })
  await fireEvent.click(confirmButton)

  expect(dummy).toBe('2024-02-02')
  expect(handleClose).toHaveBeenCalled()
})

test('onComfirm when no leading zero', async () => {
  let dummy
  const handleClose = vi.fn()
  render(
    <DatePicker
      visible={true}
      value={dayjs('2023-10-10').toDate()}
      onClose={handleClose}
      onComfirm={(value) => {
        dummy = dayjs(value).format('YYYY-MM-DD')
      }}
    />,
  )

  const [year, month, day] = screen.getAllByRole('listitem')
  await touchMoveUp(year)
  await touchMoveUp(month)
  await touchMoveUp(day)
  const confirmButton = screen.getByRole('button', { name: '确定' })
  await fireEvent.click(confirmButton)

  expect(dummy).toBe('2024-11-11')
  expect(handleClose).toHaveBeenCalled()
})

test('echo after comfirm', async () => {
  let date = '2023-11-01'
  const handleClose = vi.fn()
  const { rerender, container } = render(
    <DatePicker
      visible={true}
      value={dayjs(date).toDate()}
      onClose={handleClose}
      onComfirm={(value) => {
        date = dayjs(value).format('YYYY-MM-DD')
      }}
    />,
  )

  const day = screen.getAllByRole('listitem')[2]
  await touchMoveUp(day)
  const confirmButton = screen.getByRole('button', { name: '确定' })
  await fireEvent.click(confirmButton)

  expect(date).toBe('2023-11-02')
  expect(handleClose).toHaveBeenCalled()

  rerender(
    <DatePicker
      visible={true}
      value={dayjs(date).toDate()}
      onClose={handleClose}
      onComfirm={(value) => {
        date = dayjs(value).format('YYYY-MM-DD')
      }}
    />,
  )

  expect(container).toMatchSnapshot()
})
