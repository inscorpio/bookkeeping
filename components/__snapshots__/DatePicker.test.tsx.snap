// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`echo after comfirm 1`] = `
<div>
  <dialog
    class="w-screen max-w-none m-0 mt-auto text-stone-950 outline-none"
  >
    <div
      class="flex flex-col w-full h-[248px] py-3"
    >
      <div
        class="flex justify-between items-center px-3"
      >
        <button>
          取消
        </button>
        <button>
          确定
        </button>
      </div>
      <div
        class="flex-1 mt-2 overflow-hidden"
      >
        <div
          class="reactive h-full"
        >
          <div
            class="absolute w-full h-7 border-y border-stone-300"
          />
          <ul
            class="flex justify-center gap-6 h-full text-lg"
          >
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-280px);"
              >
                2013年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2014年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2015年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2016年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2017年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2018年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2019年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2020年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2021年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2022年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2023年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2024年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2025年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2026年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2027年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2028年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2029年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2030年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2031年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2032年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2033年
              </span>
            </li>
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-280px);"
              >
                1月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                3月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                4月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                5月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                6月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                7月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                8月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                9月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                10月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                11月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                12月
              </span>
            </li>
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-28px);"
              >
                1日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                2日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                3日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                4日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                5日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                6日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                7日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                8日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                9日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                10日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                11日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                12日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                13日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                14日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                15日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                16日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                17日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                18日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                19日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                20日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                21日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                22日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                23日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                24日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                25日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                26日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                27日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                28日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                29日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                30日
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </dialog>
</div>
`;

exports[`render 1`] = `
<div>
  <dialog
    class="w-screen max-w-none m-0 mt-auto text-stone-950 outline-none"
  >
    <div
      class="flex flex-col w-full h-[248px] py-3"
    >
      <div
        class="flex justify-between items-center px-3"
      >
        <button>
          取消
        </button>
        <button>
          确定
        </button>
      </div>
      <div
        class="flex-1 mt-2 overflow-hidden"
      >
        <div
          class="reactive h-full"
        >
          <div
            class="absolute w-full h-7 border-y border-stone-300"
          />
          <ul
            class="flex justify-center gap-6 h-full text-lg"
          >
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-280px);"
              >
                2013年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2014年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2015年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2016年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2017年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2018年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2019年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2020年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2021年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2022年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2023年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2024年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2025年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2026年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2027年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2028年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2029年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2030年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2031年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2032年
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2033年
              </span>
            </li>
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-280px);"
              >
                1月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                2月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                3月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                4月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                5月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                6月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                7月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                8月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                9月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                10月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                11月
              </span>
              <span
                style="transform: translateY(-280px);"
              >
                12月
              </span>
            </li>
            <li
              class="flex flex-col items-center overflow-hidden"
            >
              <span
                style="transform: translateY(-28px);"
              >
                1日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                2日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                3日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                4日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                5日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                6日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                7日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                8日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                9日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                10日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                11日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                12日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                13日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                14日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                15日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                16日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                17日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                18日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                19日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                20日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                21日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                22日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                23日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                24日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                25日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                26日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                27日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                28日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                29日
              </span>
              <span
                style="transform: translateY(-28px);"
              >
                30日
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </dialog>
</div>
`;
